code/modules/crafting/craft.dm

```diff
@@ -151,152 +151,125 @@
+		return ", missing tool."
+	if(!check_pathtools(user, recipe, contents))
+		return ", missing tool."
+	
+	var/list/parts = del_reqs(recipe, user)
+	if(!parts)
+		return ", missing component."
+					
+	var/atom/movable/craft_result = new recipe.result (get_turf(user.loc))
+	craft_result.CheckParts(parts, recipe)
+	if(isitem(craft_result))
+		user.put_in_hands(craft_result)
+
+	if(send_feedback)
+		SSblackbox.record_feedback("tally", "object_crafted", 1, craft_result.type)
+	return 0
+			
+
+/datum/personal_crafting/proc/del_reqs(datum/crafting_recipe/recipe, mob/user)
+	var/list/surroundings = get_environment(user)
+	var/list/parts_list_for_return = list()
+	var/list/things_for_deletion = list()
+
+	for(var/thing in recipe.reqs)
+		var/needed_amount = recipe.reqs[thing]
+		if(ispath(thing, /datum/reagent))
+			var/datum/reagent/part_reagent = locate(thing) in parts_list_for_return
+			if(!part_reagent)
+				part_reagent = new thing()
+				parts_list_for_return += part_reagent
+
+			for(var/obj/item/reagent_containers/container in surroundings)
+				var/datum/reagent/contained_reagent = container.reagents.get_reagent(thing)
+				if(!contained_reagent)
+					continue
+				
+				if(contained_reagent.volume > needed_amount)
+					things_for_deletion[container] = list(contained_reagent, needed_amount)
+					part_reagent.volume = needed_amount
+					part_reagent.data = contained_reagent.data
+					needed_amount = 0
+					break
+				else
+					needed_amount -= contained_reagent.volume
+					things_for_deletion[container] = list(contained_reagent, contained_reagent.volume)
+					part_reagent.volume += contained_reagent.volume
+					part_reagent.data += contained_reagent.data
+	
+			if(needed_amount > 0)
+				return null
+
+		else if(ispath(thing, /obj/item/stack))
+			var/obj/item/stack/part_stack = locate(thing) in parts_list_for_return
+			if(!part_stack)
+				part_stack = new thing()
+				part_stack.amount = 0
+				parts_list_for_return += part_stack
+
+			for(var/obj/item/stack/item_stack in (surroundings - things_for_deletion))
+				if(!istype(item_stack, thing))
+					continue
+				
+				if(item_stack.amount > needed_amount)
+					things_for_deletion[item_stack] = needed_amount
+					part_stack.amount += needed_amount
+					needed_amount = 0
+					break
+				else
+					needed_amount -= item_stack.amount
+					part_stack.amount += item_stack.amount
+					things_for_deletion[item_stack] = item_stack.amount
+			
+			if(needed_amount > 0)
+				return null
+
 		else
-			while(partlist[A] > 0)
-				var/atom/movable/AM = locate(A) in Deletion
-				. += AM
-				Deletion -= AM
-				partlist[A] -= 1
-	while(Deletion.len)
-		var/DL = Deletion[Deletion.len]
-		Deletion.Cut(Deletion.len)
-		qdel(DL)
+			var/atom/movable/part_atom = locate(thing) in (surroundings - parts_list_for_return)
+			if(!part_atom)
+				return null
+			parts_list_for_return += part_atom
+
+	for(var/thing in things_for_deletion)
+		if(istype(thing, /obj/item/reagent_containers))
```

Comment:
You put the reagents and the stacks into the same list, but then istype the items when iterating, to differentiate them back, and take completely separate code paths for both.
I think having two lists would be easier to follow.

Or even: handle the deletion right there in the original loop where you populate the list(s).